src/SDKs/BotService/Management.BotService/Customizations/MsaAuthenticator.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Management.BotService.Resources;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace Microsoft.Azure.Management.BotService
+{
+    /// <summary>
+    /// Obtains user level tokesn with the bot service first party app audience in order
+    /// to provision Msa Apps, which are crucial for bots to work
+    /// </summary>
+    internal class MsaAuthenticator
+    {
+        private const string BotFirstPartyAppId = "f3723d34-6ff5-4ceb-a148-d99dcd2511fc";
+        private const string AadClientId = "1950a258-227b-4e31-a9cf-717495945fc2";
+        private const string AadRedirectUri = "urn:ietf:wg:oauth:2.0:oob";
+
+        private readonly string tenantId;
+
+#if NET452
+
+        /// <summary>
+        /// Creates an instance of MsaAuthenticator
+        /// </summary>
+        public MsaAuthenticator(string tenantId)
+        {
+            this.tenantId = tenantId;
+        }
+
+#endif
+#if NETSTANDARD1_4
+
+        private readonly Action<DeviceCodeResult> deviceCodeAuthHandler;
+
+        /// <summary>
+        /// Creates an instance of MsaAuthenticator
+        /// </summary>
+        public MsaAuthenticator(string tenantId, Action<DeviceCodeResult> deviceCodeAuthHandler)
+        {
+            this.tenantId = tenantId;
+            this.deviceCodeAuthHandler = deviceCodeAuthHandler;
+        }
+
+#endif
+
+        /// <summary>
+        /// Acquires a user token with the bot service first party app as audience
+        /// </summary>
+        /// <returns></returns>
+        public async Task<AuthenticationResult> AcquireTokenAsync()
+        {
+            var authority = $"https://login.windows.net/{tenantId}";
+            var context = new AuthenticationContext(
+                authority: authority,
+                validateAuthority: true);
+            
+            SynchronizationContext.SetSynchronizationContext(new SynchronizationContext());
+
+#if NET452
```

Comment:
I wanted to replace all occurrences of NET452 with FullNetFx and replace NETSTANDARD1_4 with !FullNetFx. However, this doesn't compile for me. The problem is that things such as NETSTANDARD1_4  don't allow multitargetting. What would be the right conditions here? Or alternatively, what are the targets for which we publish the nuget so I can test them all? Thanks!

Reply:
@carlosscastro what build errors are you getting.
Also can you clarify what you mean when you say, "NETSTANDARD1_4 don't allow multi-targeting"

Reply:
Well, NETSTANDARD1_4 and NET452 works perfect for me, but I'm not sure what are the targets you build against. If those 2 are enough then we should be fine.